contracts/contracts/coordination/Coordinator.sol

```diff
@@ -124,6 +126,11 @@ contract Coordinator is AccessControlDefaultAdminRules {
         _setRoleAdmin(INITIATOR_ROLE, bytes32(0));
     }
 
+    function setProviderPublicKey(BLS12381.G2Point calldata publicKey) external {
```

Comment:
does this need some access control?

Reply:
Good question. One can only set a public key for themselves so I assumed it's one of the cases when we don't necessarily need AC

Reply:
Agreed

Reply:
Would we ever consider the case that Ursulas would have a different key per ritual? Is there any advantage to that?

Does this model not allow the key to change at all...

Reply:
> Would we ever consider the case that Ursulas would have a different key per ritual? Is there any advantage to that?

I think so, but IIRC, we currently only have long-term Ferveo keys, so for the moment this works.

> Does this model not allow the key to change at all... eg. if it was compromised etc.

Isn't the opposite? This model allow to set the key at any time.

Reply:
OK I see what you mean. Well, the very basic requirement here is to have a minimal form of commitment and retrievability of the Ferveo key, since for the moment we are not going to be able to do any on-chain verification. This means that, even if keys are changed at a given time, there's always a record of the previous keys (not accessible via contract state, but accessible through the TX hash or historical data). For our purposes, I think it could suffice. 

Now, if we want to improve this in line with your comment, I think it wouldn't take much effort. We could do something like, e.g., an append-only list where for each node the Coordinator stores a pair of `(public_key, start_ritual)`, so in the optimistic case this list is of length 1.

Reply:
I think for the purpose of local verification emitting `ParticipantPublicKeySet(public_key)` event would also be acceptable. Local verification should be performed at least once per ritual and searching through events shouldn't be too much of a hassle in this case. 

For on-chain verification mapping is sufficient and no access to historic public key values is needed.

Reply:
> This means that, even if keys are changed at a given time, there's always a record of the previous keys (not accessible via contract state, but accessible through the TX hash or historical data).

Right.

Nice thought! If we accept that sleuthing blockchain information is acceptable for verification an event could be a good solution 👍

Reply:
> Would we ever consider the case that Ursulas would have a different key per ritual? Is there any advantage to that?
> 
> Does this model not allow the key to change at all... eg. if it was compromised etc.

This is also my impression, there's no harm in supporting ferveo public keys in the Ritual struct despite ursula's only using a long-term public key at present. The downstream logic is easier to understand and implement imo.

Reply:
> ParticipantPublicKeySet(public_key) event would also be acceptable

Rethinking this I'm not sure that this is really that feasible. For a Cohort say of size 16-of-30, how feasible is it to possibly look for / process these events. I'm thinking with respect to @piotr-roslaniec 's local verification PR, that would need a lot more logic to accommodate these events.

> This is also my impression, there's no harm in supporting ferveo public keys in the Ritual struct

This was my first inclination as well but I can see the argument that it involves extra contract storage space and additional gas costs which may be mostly wasted *if* the key never changes.

> e.g., an append-only list where for each node the Coordinator stores a pair of (public_key, start_ritual), so in the optimistic case this list is of length 1.

Maybe we're back to @cygnusv 's solution and perhaps if we use this type of storage, then we provide a getter function in the contract that takes a ritual number and a provider address, and returns the relevant key...maybe...? (cc @vzotova )

Reply:
> For a Cohort say of size 16-of-30, how feasible is it to possibly look for / process these events

I don't think it's too bad: Knowing the ritual timestamp, for each participant we need to find the "last-before" event. I don't see this as an issue because I assumed that participants wouldn't change their keys very often and that the filtering is done over JSON-RPC anyway.

`(public_key, start_ritual)` sounds good to me.

Reply:
@cygnusv Does this implementation sketch make sense?

```
    mapping(address => KeyRecord[]) keysHistory;
    event ParticipantPublicKeySet(address indexed participant, BLS12381.G2Point publicKey);
    
    function setPublicKey(bytes calldata _publicKey) public {
        BLS12381.G2Point memory publicKey = BLS12381.bytesToG2Point(_publicKey);
        uint startRitual = block.timestamp;
        
        KeyRecord memory newRecord = KeyRecord(publicKey, startRitual);
        keysHistory[msg.sender].push(newRecord);
        
        emit ParticipantPublicKeySet(msg.sender, publicKey);
    }

    function getPreviousKey(address _address, uint _ritual) public view returns (BLS12381.G2Point memory) {
        KeyRecord[] storage userHistory = keysHistory[_address];
        require(userHistory.length > 0, 'No keys found for the provided address');

        for(uint i = userHistory.length - 1; i >= 0; i--) {
            if(userHistory[i].startRitual <= _ritual) return userHistory[i].publicKey;
        }
        
        revert('No keys found prior to the provided ritual');
    }
    ```

Reply:
Looks good in genera, but some comments: 
* Why `uint startRitual = block.timestamp;`? Can we just do `rituals.length` which is the ID of the next ritual?
* `msg.sender` is the operator, so we need to resolve the provider. 
* We could remove the `require(participantHistory.length > 0, 'No keys found for the provided address');`